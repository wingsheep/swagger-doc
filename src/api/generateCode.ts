import path from 'path'
import fs from 'fs'
import _ from 'lodash'

import { generateApi } from 'swagger-typescript-api'
import type { Spec } from 'swagger-schema-official'

const FILE_PREFIX = `/* eslint-disable */
/* tslint:disable */

`

export function generateCode(data, floderPath, toJs = false) {
  return new Promise((resolve) => {
    generateApi({
      name: 'api.ts',
      output: false,
      spec: data as Spec,
      httpClientType: 'axios',
      generateClient: true,
      singleHttpClient: false,
      moduleNameFirstTag: false,
      modular: true,
      toJS: toJs,
      templates: path.resolve(process.cwd(), './src/__templates__/modular'),
    }).then(({ files }) => {
      const dir = path.resolve(floderPath, './__generated__')
      if (!fs.existsSync(dir))
        fs.mkdirSync(dir, { recursive: true })

      files.forEach(({ content, name }) => {
        fs.writeFileSync(path.resolve(dir, `./${name}`), `${FILE_PREFIX}${content}`, _.noop)
      })
      resolve(true)
    })
      .catch(e => console.error(e))
  })
}

// export function generateTemplate() {
//   return new Promise((resolve) => {
//     const data = generateTemplates({
//       cleanOutput: false,
//       output: path.resolve(process.cwd(), './src/__templates__'),
//       httpClientType: 'axios',
//       modular: false,
//       silent: false,
//     })
//     resolve(data)
//   })
// }
