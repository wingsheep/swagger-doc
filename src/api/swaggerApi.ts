import SwaggerParser from '@apidevtools/swagger-parser'
import type { OpenAPIV2, OpenAPIV3 } from 'openapi-types'

interface TagTree {
  name: string
  description: string
  children: any[]
}
export class SwaggerApi {
  public static currentDoc: SwaggerApi | undefined
  /** Swagger json data */
  public static docJson: OpenAPIV2.Document
  /** Swagger tag tree */
  public static tagTree: TagTree[] | undefined
  /**
   * Get swagger doc json
   * @param {string} url https://dev-apisix.hgj.com/sonny-vehicle-account-book/v2/api-docs
   * @memberof SwaggerApi
   */
  public static async getDocJson(url: string | OpenAPIV2.Document) {
    try {
      const api = await SwaggerParser.dereference(url) as OpenAPIV2.Document
      SwaggerApi.currentDoc = new SwaggerApi(api)
    }
    catch (error) {
      SwaggerApi.tagTree = undefined
    }
  }

  /**
   * @description GET api detail data
   * @static
   * @param {string} path api url
   * @param {string} method api method
   * @return {*}
   * @memberof SwaggerApi
   */
  public static getApiDetail(path: string, method: string) {
    const { paths } = SwaggerApi.docJson
    const pathValue = paths[path]
    const apiDetail = pathValue && (pathValue as OpenAPIV2.PathsObject)[method] as OpenAPIV2.OperationObject
    Object.keys(apiDetail.responses).forEach((code) => {
      const responseValue = apiDetail.responses[code] as (OpenAPIV2.ResponseObject | OpenAPIV2.ScopesObject)
      if (responseValue && responseValue?.schema)
        responseValue.examples = getSchemaExample(responseValue?.schema)
    })
    apiDetail.parameters?.forEach((parameter: any) => {
      if (parameter.schema)
        parameter.examples = getSchemaExample(parameter.schema)
    })

    return apiDetail
  }

  private constructor(data: OpenAPIV2.Document) {
    SwaggerApi.docJson = data
    SwaggerApi.tagTree = this._getTagTree()
  }

  /**
   * @description GET tag tree
   * @private
   * @return {*}
   * @memberof SwaggerApi
   */
  private _getTagTree() {
    const { paths, tags } = SwaggerApi.docJson
    const apiTreeList: TagTree[] = []
    tags?.forEach((tag) => {
      const treeItem = {
        name: tag.name,
        description: '',
        children: [] as any,
      }
      Object.keys(paths).forEach((path) => {
        const pathValue = paths[path]
        Object.keys(pathValue).forEach((method) => {
          const methodValue = (pathValue as OpenAPIV2.PathsObject)[method] as OpenAPIV2.OperationObject
          if (methodValue?.tags && methodValue.tags.includes(tag.name)) {
            treeItem.children.push({
              name: `${method.toLocaleUpperCase()} ${methodValue.summary}`,
              id: methodValue.operationId,
              api: {
                method,
                tags: methodValue.tags,
                path,
              },
            })
          }
        })
      })
      treeItem.description = `(${treeItem.children.length})`
      apiTreeList.push(treeItem)
    })
    return apiTreeList
  }

  /**
   * @description GET api schemas
   * @private
   * @memberof SwaggerApi
   */
  private _getApiSchemas() {
    const { definitions } = SwaggerApi.docJson
    const apiSchemas: any = []
    if (definitions) {
      Object.keys(definitions).forEach((definition) => {
        const definitionValue = definitions[definition]
        const apiSchemasItem = {
          name: definition,
          jsonSchema: definitionValue,
        }
        apiSchemas.push(apiSchemasItem)
      })
    }
  }
}

/**
 * @description Get response example
 * @param {*} schema
 * @return {*}
 */
function getSchemaExample(schema: any) {
  const example = schema.example || generateExample(schema)
  return example
}

/**
 * @description Generate example data
 * @param {*} schema
 * @return {*}  {unknown}
 */
function generateExample(schema: any): unknown {
  const type = schema.type
  if (type === 'object') {
    const properties = schema.properties
    const example: any = {}
    for (const key in properties) {
      const prop = properties[key]
      example[key] = generateExample(prop)
    }
    return example
  }
  else if (type === 'array') {
    const items = schema.items
    return [generateExample(items)]
  }
  else {
    return schema.example || getDefaultByType(type)
  }
}
/**
 * @description Get default by type
 * @param {OpenAPIV3.NonArraySchemaObjectType} type
 * @return {*}  {unknown}
 */
function getDefaultByType(type: OpenAPIV3.NonArraySchemaObjectType): unknown {
  let result
  switch (type) {
    case 'string':
      result = 'default string value'
      break
    case 'integer':
    case 'number':
      result = 0
      break
    case 'boolean':
      result = true
      break
    default:
      result = null
      break
  }
  return result
}
