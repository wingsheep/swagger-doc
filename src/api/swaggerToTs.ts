import path from 'path'
import { generateApi, generateTemplates } from 'swagger-typescript-api'
import type { Spec } from 'swagger-schema-official'
import swaggerJSON from '../../backend/mock/swagger.json'
export function generate() {
  return new Promise((resolve) => {
    generateApi({
      name: 'MySuperbApi.ts',
      // path: '/home/monthlySummary',
      // set to `false` to prevent the tool from writing to disk
      output: path.resolve(process.cwd(), './src/__generated__'),
      url: '/home/monthlySummary',
      spec: swaggerJSON as Spec,
      httpClientType: 'axios',
      generateClient: true,
      singleHttpClient: true,
      templates: path.resolve(process.cwd(), '../src/__templates__'),
    }).then(({ files, configuration }) => {
      resolve([])
    })
      .catch(e => console.error(e))
  })
}

export function generateTemplate() {
  return new Promise((resolve) => {
    const data = generateTemplates({
      cleanOutput: false,
      output: path.resolve(process.cwd(), './src/__templates__'),
      httpClientType: 'axios',
      modular: false,
      silent: false,
    })
    resolve(data)
  })
}
