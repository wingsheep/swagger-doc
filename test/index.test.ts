import { describe, expect, it } from 'vitest'
import { SwaggerApi } from '../src/api/swaggerApi'
describe('should', () => {
  it('exported', async () => {
    // const url = 'https://beta-apisix.hgj.com/sonny-vehicle-account-book/v2/api-docs'
    const mockUrl = 'http://127.0.0.1:3000/v2/example'
    await SwaggerApi.getDocJson(mockUrl)
    // await expect(SwaggerApi.docJson).toMatchInlineSnapshot()
    // await expect(SwaggerApi.tagTree).toMatchInlineSnapshot()
    await expect(SwaggerApi.tagList).toMatchInlineSnapshot(`
      [
        {
          "api": {
            "method": "get",
            "path": "/profit/analysis/customer",
            "tags": [
              "利润分析",
            ],
          },
          "id": "customerProfitAnalysisUsingGET",
          "label": "GET 客户利润列表",
        },
        {
          "api": {
            "method": "get",
            "path": "/profit/analysis/income",
            "tags": [
              "利润分析",
            ],
          },
          "id": "incomeProfitAnalysisUsingGET",
          "label": "GET 收入支出分析",
        },
        {
          "api": {
            "method": "get",
            "path": "/profit/analysis/line",
            "tags": [
              "利润分析",
            ],
          },
          "id": "lineProfitAnalysisUsingGET",
          "label": "GET 线路利润列表",
        },
        {
          "api": {
            "method": "get",
            "path": "/profit/analysis/monthlySummary",
            "tags": [
              "利润分析",
            ],
          },
          "id": "detailUsingGET_1",
          "label": "GET 月度汇总",
        },
        {
          "api": {
            "method": "get",
            "path": "/profit/analysis/task",
            "tags": [
              "利润分析",
            ],
          },
          "id": "taskProfitAnalysisUsingGET",
          "label": "GET 任务利润列表",
        },
        {
          "api": {
            "method": "post",
            "path": "/address/delete",
            "tags": [
              "地址簿相关",
            ],
          },
          "id": "deleteUserAddressUsingPOST",
          "label": "POST 删除用户地址簿",
        },
        {
          "api": {
            "method": "get",
            "path": "/address/list",
            "tags": [
              "地址簿相关",
            ],
          },
          "id": "listUsingGET_2",
          "label": "GET 查看用户地址簿",
        },
        {
          "api": {
            "method": "get",
            "path": "/address/vagueQuery",
            "tags": [
              "地址簿相关",
            ],
          },
          "id": "vagueQueryUsingGET",
          "label": "GET 模糊搜索",
        },
        {
          "api": {
            "method": "post",
            "path": "/customer/add",
            "tags": [
              "客户相关",
            ],
          },
          "id": "addUsingPOST_2",
          "label": "POST 新增客户",
        },
        {
          "api": {
            "method": "post",
            "path": "/customer/delete",
            "tags": [
              "客户相关",
            ],
          },
          "id": "deleteUsingPOST_2",
          "label": "POST 删除客户",
        },
        {
          "api": {
            "method": "get",
            "path": "/customer/list",
            "tags": [
              "客户相关",
            ],
          },
          "id": "listUsingGET_3",
          "label": "GET 查看用户客户列表",
        },
        {
          "api": {
            "method": "post",
            "path": "/accountCategory/add",
            "tags": [
              "收支分类相关",
            ],
          },
          "id": "addAccountCategoryUsingPOST",
          "label": "POST 新增账单类型",
        },
        {
          "api": {
            "method": "post",
            "path": "/accountCategory/delete",
            "tags": [
              "收支分类相关",
            ],
          },
          "id": "deleteAccountCategoryUsingPOST",
          "label": "POST 删除账单类型",
        },
        {
          "api": {
            "method": "get",
            "path": "/accountCategory/list",
            "tags": [
              "收支分类相关",
            ],
          },
          "id": "accountCategoryListUsingGET",
          "label": "GET 账单类型分类列表",
        },
        {
          "api": {
            "method": "post",
            "path": "/accountBook/add",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "addUsingPOST",
          "label": "POST 记支出/收入",
        },
        {
          "api": {
            "method": "post",
            "path": "/accountBook/delete",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "deleteUsingPOST",
          "label": "POST 删除账单流水",
        },
        {
          "api": {
            "method": "get",
            "path": "/accountBook/detail",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "detailUsingGET",
          "label": "GET 流水详情",
        },
        {
          "api": {
            "method": "get",
            "path": "/accountBook/list",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "listUsingGET",
          "label": "GET 查询流水记录",
        },
        {
          "api": {
            "method": "get",
            "path": "/accountBook/listByTaskId",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "listByTaskIdUsingGET",
          "label": "GET 根据运单任务id查询流水记录",
        },
        {
          "api": {
            "method": "get",
            "path": "/accountBook/listByYearMonth",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "listByYearMonthUsingGET",
          "label": "GET 根据年月日期 查询列表",
        },
        {
          "api": {
            "method": "post",
            "path": "/accountBook/update",
            "tags": [
              "账单流水相关",
            ],
          },
          "id": "updateUsingPOST",
          "label": "POST 编辑支出/收入",
        },
        {
          "api": {
            "method": "post",
            "path": "/freightTask/add",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "addUsingPOST_1",
          "label": "POST 新增运单",
        },
        {
          "api": {
            "method": "post",
            "path": "/freightTask/delete",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "deleteUsingPOST_1",
          "label": "POST 删除运单",
        },
        {
          "api": {
            "method": "get",
            "path": "/freightTask/getDetail",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "getDetailUsingGET",
          "label": "GET 获取任务详情",
        },
        {
          "api": {
            "method": "get",
            "path": "/freightTask/getQuery",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "getQueryUsingGET",
          "label": "GET 获取筛选条件",
        },
        {
          "api": {
            "method": "get",
            "path": "/freightTask/list",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "listUsingGET_1",
          "label": "GET 运单列表",
        },
        {
          "api": {
            "method": "get",
            "path": "/freightTask/listFromAnalysis",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "listFromAnalysisUsingGET",
          "label": "GET 从利润分析跳转运单列表",
        },
        {
          "api": {
            "method": "post",
            "path": "/freightTask/update",
            "tags": [
              "运单任务相关",
            ],
          },
          "id": "updateUsingPOST_1",
          "label": "POST 编辑运单",
        },
        {
          "api": {
            "method": "get",
            "path": "/home/monthlySummary",
            "tags": [
              "首页相关",
            ],
          },
          "id": "monthlySummaryUsingGET",
          "label": "GET 首页月度统计",
        },
      ]
    `)
    // expect(SwaggerApi.getApiDetail('/accountBook/add', 'post')).toMatchInlineSnapshot()
    // expect(SwaggerApi.getApiDetail('/freightTask/add', 'post')).toMatchInlineSnapshot()
  })
})
