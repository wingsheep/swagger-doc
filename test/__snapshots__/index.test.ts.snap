// Vitest Snapshot v1

exports[`should > exported 1`] = `
{
  "consumes": [
    "application/json",
  ],
  "operationId": "addUsingPOST",
  "parameters": [
    {
      "definition": "AccountBookAddRequest",
      "description": "request",
      "in": "body",
      "jsonSchema": {
        "properties": {
          "amount": {
            "allowEmptyValue": false,
            "description": "金额 (单位分)",
            "example": 1000,
            "format": "int64",
            "type": "integer",
          },
          "categoryId": {
            "allowEmptyValue": false,
            "description": "类型Id",
            "example": 1567054180388524000,
            "format": "int64",
            "type": "integer",
          },
          "categoryName": {
            "allowEmptyValue": false,
            "description": "类型名称",
            "example": "加油",
            "type": "string",
          },
          "occurrenceTime": {
            "allowEmptyValue": false,
            "description": "时间",
            "example": "yyyy-MM-dd",
            "format": "date",
            "type": "string",
          },
          "remarks": {
            "allowEmptyValue": false,
            "description": "备注",
            "example": "买的燃油宝",
            "type": "string",
          },
          "taskId": {
            "allowEmptyValue": false,
            "description": "运单id",
            "example": 1750148943159,
            "format": "int64",
            "type": "integer",
          },
          "type": {
            "allowEmptyValue": false,
            "description": "类别 0:收入 1:支出",
            "example": 1,
            "format": "int32",
            "type": "integer",
          },
        },
        "title": "AccountBookAddRequest",
        "type": "object",
      },
      "name": "request",
      "required": true,
      "schema": {
        "$ref": "#/definitions/AccountBookAddRequest",
      },
    },
  ],
  "produces": [
    "*/*",
  ],
  "responses": {
    "200": {
      "description": "OK",
    },
    "201": {
      "description": "Created",
    },
    "401": {
      "description": "Unauthorized",
    },
    "403": {
      "description": "Forbidden",
    },
    "404": {
      "description": "Not Found",
    },
  },
  "summary": "记支出/收入",
  "tags": [
    "账单流水相关",
  ],
}
`;
